### ISSUE TITLE
 Book cover in Kobo didn't get sync with library change

### ISSUE THREAD

**Issue by fancl20** [janeczku/calibre-web#2509]
**Describe the bug/problem**
The book cover in Kobo didn't get updated after the cover in the library changed.

**To Reproduce**
1. Add book and its cover
2. Sync with Kobo
3. Change the cover
4.  Sync with Kobo, cover won't update.

Tried force sync all.

**Expected behavior**
Cover updated

**Comment by OzzieIsaacs** [comment]
This is indeed the case, changing title, author or something else is also not working. I haven‘t found a solution for this yet

**Comment by fancl20** [comment]
Is this a limitation on Kobo side? I also can't find a way to force kobo remove the local metadata even with "repair account". Although that do download the cover if it's missing.

**Comment by OzzieIsaacs** [comment]
Yeah, kobo stores the metadata and if you change it afterwards kobo doesn‘t care and uses the old ones

**Comment by fancl20** [comment]
Maybe we can calculate the book id (for kobo) by hashing metadata? So kobo will simply treat is as a different book if metadata changed.

**Comment by Fosnine** [comment]
Regarding the cover:
It seems that a workaround is to connect the Kobo to a PC and delete the content of the relevant book from .kobo-images.
I added a couple of books before making sure the covers were good and was stuck with the same issue you described.
I deleted all the folders in .kobo-images and did User->Force Full Kobo Sync in calibre-web (not sure whether the last step is necessary but I think it was in my case). Next time you sync, Kobo requests all the covers again from calibre-web.
There's a separate issue regarding the resolution but #2817 has been opened for that.

This won't solve the metadata issue but at least you get prettier covers. 😃

**Comment by AsherMaximum** [comment]
I have a PR that covers a few changes to the sync process - deleting books from a kobo synced shelf, and updating the metadata.

Covers weren't getting updated, but I just added a commit that adds that as well.

The issue was that the coverImageId that calibre-web sends is just the UUID of the book. So when the cover changed, the URL for the coverImageId was exactly the same as before, and the Kobo device cached it.
I added the epoch timestamp of the date the thumbnail for the cover was created to the coverImageId, so now whenever the cover is changed and refreshThumbnailCache is run, the cover will be re-downloaded by the Kobo.

Note that for me at least (on a Kobo Libra 2), the cover isn't re-downloaded after I sync until I navigate away from the Books page and go back to it.

**Comment by AsherMaximum** [comment]
That was easier than I thought - the cover will now work without having refresh thumbnail cache turned on.

Additionally, if you do have refresh thumbnails on, they will be refreshed at the start of a sync automatically

**Comment by new-usemame** [comment]
If anyone here is still seeing stale covers after a library cover swap, the community-maintained CWA build at [Calibre-Web-NextGen](https://github.com/new-usemame/Calibre-Web-NextGen) handles this since v4.0.21.

The change is in `cps/kobo.py:_get_cover_image_id`: when the device asks "what's this book's cover ID," the response now includes the cover.jpg `mtime`. Replacing the cover updates the file's mtime, which flips the CoverImageId, which makes the Kobo treat the cover as new and re-download it. No force-sync, no toggle dance — change the cover, next sync picks it up.

`docker pull ghcr.io/new-usemame/calibre-web-nextgen:latest`. Heads-up: this build is CWA-derived, so it bakes in auto-ingest, kepubify, and KOReader sync on top of the calibre-web core. If you're on stock CW and don't want those, it's not the right fit.
